Implement video filtering with libavfilter

KingOfTheBlues opened the pull request:
Adds bindings for libavfilter attempting to follow established patterns where possible. Audio filters to be added in later.

The implementation is based on the ffmpeg example: https://www.ffmpeg.org/doxygen/trunk/filtering_8c-source.html

Thanks in advance for your time considering this PR and providing any feedback.

operutka commented:
Hi @KingOfTheBlues and thank you for the PR. I've checked it just briefly and it's looking good. I'll have to find some time to look into it in detail though.

Until then, please put `avfilter` behind a feature gate. Without it, we could be breaking existing builds by introducing a new link time dependency which may not be available (especially if a custom build of FFmpeg is being used).

operutka commented on ac-ffmpeg/src/codec/video/filter.c:
```diff
@@ -0,0 +1,139 @@
+    inputs->pad_idx = 0;
+    inputs->next = NULL;
+
+    ret = avfilter_graph_parse_ptr(filter_graph, filters_descr, &inputs, &outputs, NULL);
+    if (ret < 0) {
+        return ret;
+    }
+
+    ret = avfilter_graph_config(filter_graph, NULL);
+    if (ret < 0) {
+        return ret;
+    }
+
+    return ret;
```
Based on the FFmpeg documentation and the example [here](http://ffmpeg.org/doxygen/5.1/filtering_video_8c-example.html#_a32), the `inputs` and `outputs` should be deallocated.

```suggestion
    if ((ret = avfilter_graph_parse_ptr(filter_graph, filters_descr, &inputs, &outputs, NULL)) < 0) {
        goto end;
    }

    if ((ret = avfilter_graph_config(filter_graph, NULL)) < 0) {
        goto end;
    }

end:
    avfilter_inout_free(&inputs);
    avfilter_inout_free(&outputs);

    return ret;
```

operutka commented on ac-ffmpeg/src/codec/video/filter.c:
```diff
@@ -0,0 +1,139 @@
+}
+
+int ffw_filtergraph_push_frame(AVFilterContext* context, AVFrame* frame) {
+    int ret = av_buffersrc_add_frame(context, frame);
+
+    if (ret == 0 || ret == AVERROR_EOF) {
+        return 1;
+    }
+    else if (ret == AVERROR(EAGAIN)) {
+        return 0;
+    }
+
+    return ret;
```
I'm not sure if this is correct. The FFmpeg documentation does not say that `av_buffersrc_add_frame` would behave the same way as `avcodec_send_frame`. Is it possible that pushing one frame to a filter input can generate multiple frames at the filter output? Do we also have to consume all output frames before pushing the next input frame (i.e. is it possible that `av_buffersrc_add_frame` would return `AVERROR(EAGAIN)`)?

KingOfTheBlues replied:
thought I had addressed this one, actually I reverted the change for now as applying the respective change in the C code yields "all frames must be consumed before flushing" from CodecError

operutka commented on ac-ffmpeg/src/codec/video/filter.rs:
```diff
@@ -0,0 +1,232 @@
+        };
+        if ret < 0 {
+            return Err(Error::from_raw_error_code(ret));
+        } else if self.ptr.is_null() {
+            return Err(Error::new("unable to allocate a filtergraph"));
+        }
```
This check probably isn't necessary.

KingOfTheBlues commented:
thanks so much @operutka for the awesome feedback! The code is now greatly simplified and I really appreciate the help here. I believe I have addressed all your comments now
